src/main/java/jimmy/mcgymmy/model/ModelManager.java

```diff
@@ -98,20 +102,43 @@ public boolean hasFood(Food food) {
 
     @Override
     public void deleteFood(Index index) {
+        addCurrentStateToHistory();
         mcGymmy.removeFood(index);
     }
 
     @Override
     public void addFood(Food food) {
+        addCurrentStateToHistory();
         mcGymmy.addFood(food);
         updateFilteredFoodList(PREDICATE_SHOW_ALL_FOODS);
     }
 
     @Override
     public void setFood(Index index, Food editedFood) {
         CollectionUtil.requireAllNonNull(index, editedFood);
-
+        addCurrentStateToHistory();
         mcGymmy.setFood(index, editedFood);
+        updateFilteredFoodList(PREDICATE_SHOW_ALL_FOODS);
+    }
+
+    @Override
+    public boolean canUndo() {
+        return !mcGymmyStack.empty();
+    }
+
+    /**
+     * Undo the previous change to mcGymmy
+     */
+    @Override
+    public void undo() {
+        if (canUndo()) {
+            mcGymmy.resetData(mcGymmyStack.pop());
+            updateFilteredFoodList(PREDICATE_SHOW_ALL_FOODS);
+        }
```

Comment:
Can add some assertions here regarding calling the function when it is unable to undo

Reply:
I've added one assertion in the latest commit